### PR TITLE
8366421: ModifiedUtf.utfLen may overflow for giant string

### PR DESCRIPTION
**Pull request by hgqxjj**
Please review this patch.

**Description:**

Currently, ModifiedUtf.utfLen returns a signed int. For very large strings, this may overflow and produce negative values, leading to incorrect behavior in code that relies on the UTF length. This patch changes the return type to long, which fully resolves the issue and allows safe handling of giant strings.

**Test:**

GHA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8366421](https://bugs.openjdk.org/browse/JDK-8366421): ModifiedUtf.utfLen may overflow for giant string (**Enhancement** - P5)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/27285/head:pull/27285` \
`$ git checkout pull/27285`

Update a local copy of the PR: \
`$ git checkout pull/27285` \
`$ git pull https://git.openjdk.org/jdk.git pull/27285/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 27285`

View PR using the GUI difftool: \
`$ git pr show -t 27285`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/27285.diff">https://git.openjdk.org/jdk/pull/27285.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/27285#issuecomment-3290865572)
</details>

### CODE REVIEW

**Comment by hgqxjj** [comment]
Hi @RogerRiggs @liach ,

Could you kindly take a look at this PR when you have time?  Thanks!

**Comment by RogerRiggs** [comment]
Can you add a test of the maximum length UTF-8 encoded string. 
That would be a string of Integer.MAX_VALUE/2 characters that were > 0xff.
It will likely have to write it to a file and read it back, ByteArrayIn/OutStream wouldn't be big enough.

**Comment by liach** [comment]
Do you think we should use `> 65535L` or `ExactConversionsSupport.isLongToCharExact`?

**Comment by hgqxjj** [comment]
Hi @RogerRiggs @liach 
Thanks for the suggestion. 
Creating a string of Integer.MAX_VALUE/2 characters would require enormous memory, even using a file, since the JVM still needs to hold the string content in memory when reading it back.
Instead, i used a small string chunk with 1-, 2-, and 3-byte UTF-8 characters and repeatedly called ModifiedUtf.utfLen() in a loop, accumulating the total in a long. This safely simulates a total length exceeding Integer.MAX_VALUE and verifies that the change to long prevents overflow.
Could you please take another look when you have time ? Thanks!

**Comment by liach** [comment]
> This safely simulates a total length exceeding Integer.MAX_VALUE and verifies that the change to long prevents overflow.

Can you try derive a **regression** test that fails on mainline and passes with your fix? You can generate a string with `chunk.repeat(iterations)` and run `ModifiedUtf.utfLen` on it.

File: test/jdk/jdk/internal/util/TestUtfLen.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        int iterations = (Integer.MAX_VALUE / 6) + 1;
+        long total = 0L;
+        for (int i = 0; i < iterations; i++) {
+            total += ModifiedUtf.utfLen(chunk, 0);
+        }
+        long expected = perChunkLen * iterations;
+        if (total != expected) {
+            throw new RuntimeException("Expected total=" + expected + " but got " + total);
+        }
```
Comment (RogerRiggs) [inline test/jdk/jdk/internal/util/TestUtfLen.java]: This misses the point of testing utfLen; it is always testing the same string testing only the inner path through the code.
The PR changes the callers of utfLen and they should be tested, not just the internal function.
The test really does need to create the large string and pass it to the APIs that have been changed.

File: test/jdk/jdk/internal/util/TestUtfLen.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        long perChunkLen = ModifiedUtf.utfLen(chunk, 0);
+        if (perChunkLen != 6L) {
+            throw new RuntimeException("Expected perChunkLen=6 but got " + perChunkLen);
+        }
+
+        int iterations = (Integer.MAX_VALUE / 6) + 1;
+        long total = 0L;
+        for (int i = 0; i < iterations; i++) {
+            total += ModifiedUtf.utfLen(chunk, 0);
+        }
```
Comment (liach) [inline test/jdk/jdk/internal/util/TestUtfLen.java]:
```suggestion
        long total = ModifiedUtf.utfLen(chunk.repeat(iterations), 0);
```

File: test/jdk/jdk/internal/util/TestUtfLen.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * @test
+ * @bug 8366421
+ * @summary Test for ModifiedUtf.utfLen() return type change from int to long to avoid overflow
+ * @modules java.base/jdk.internal.util
+ * @run main TestUtfLen
+ */
+
+import jdk.internal.util.ModifiedUtf;
+
+public class TestUtfLen {
```
Comment (liach) [inline test/jdk/jdk/internal/util/TestUtfLen.java]: Please rename this to `ModifiedUtfTest`, convert to junit, and turn this into one case for overflow of `utfLen`, like `testUtfLenOverflow`, to align with other tests in the same directory.

Core libraries usually use JUnit tests instead of main tests like runtime unless there are sensitive stuff that need flagless or would be perturbed by JUnit or other framework initialization.

Comment (liach) [reply test/jdk/jdk/internal/util/TestUtfLen.java]: Correction: this test will allocate a 2GB string, so we might still have this test itself as a main and focus on `utfLen` overflow alone, and add something like `@requires os.maxMemory >= 4g`

File: test/jdk/jdk/internal/util/TestUtfLen.java (hunk 2 above)
Comment (hgqxjj) [reply test/jdk/jdk/internal/util/TestUtfLen.java]: String.repeat() cannot generate a string whose total length exceeds Integer.MAX_VALUE due to internal limits. That’s why I used a small chunk and accumulated UTF-8 length in a loop.It seems that the String type cannot hold a string whose length exceeds Integer.MAX_VALUE.
https://github.com/openjdk/jdk/blob/e3a4c28409ac62feee9efe069e3a3482e7e2cdd2/src/java.base/share/classes/java/lang/String.java#L4875

Comment (liach) [reply test/jdk/jdk/internal/util/TestUtfLen.java]:
```
jshell --add-exports java.base/jdk.internal.util=ALL-UNNAMED
|  Welcome to JShell -- Version 24
|  For an introduction type: /help intro

jshell> import jdk.internal.util.ModifiedUtf;

jshell> var s = "\u0100\u0100\u2600".repeat(Integer.MAX_VALUE / 6 - 1);
s ==> "???????????????????????????????????????????????? ... ?????????????????????????"

jshell> ModifiedUtf.utfLen(s)
|  Error:
|  method utfLen in class jdk.internal.util.ModifiedUtf cannot be applied to given types;
|    required: java.lang.String,int
|    found:    java.lang.String
|    reason: actual and formal argument lists differ in length
|  ModifiedUtf.utfLen(s)
|  ^----------------^

jshell> ModifiedUtf.utfLen(s, 0)
$3 ==> -1789569716
```

You can construct such a string if the number of bytes in the Modified UTF 8 form is more than the number of bytes in UTF16 form, such as if you use all 3-byte characters.

**Comment by dholmes-ora** [comment]
> Creating a string of Integer.MAX_VALUE/2 characters would require enormous memory

See test/hotspot/jtreg/runtime/jni/checked/TestLargeUTF8Length.java

File: test/jdk/jdk/internal/util/TestUtfLen.java (hunk 2 above)
Comment (hgqxjj) [reply test/jdk/jdk/internal/util/TestUtfLen.java]: Got it!

**Comment by hgqxjj** [comment]
@RogerRiggs @liach @dholmes-ora 

Thank you for your suggestion! 
I’ve optimized the regression test as requested — could you please take another look when you have time? Thanks!

File: test/jdk/jdk/internal/util/ModifiedUtfTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+     */
+    int lengthFieldSize = 8;
+    // Offset to UTF length field: 2 bytes STREAM_MAGIC + 2 bytes STREAM_VERSION + 5 bytes block data header
+    int lengthFieldOffset = 9;
+    int headerSize = 20; // greater than lengthFieldSize + lengthFieldOffset
+    HeaderCaptureOutputStream headerOut = new HeaderCaptureOutputStream(headerSize);
+    try (ObjectOutputStream objOut = new ObjectOutputStream(headerOut)) {
+      objOut.writeUTF(largeString);
+    } catch (Exception e) {
+    } finally {
```
Comment (liach) [inline test/jdk/jdk/internal/util/ModifiedUtfTest.java]: You should rethrow the exception here in case of failure instead of truncating, and you no longer need the finally block.

File: test/jdk/jdk/internal/util/ModifiedUtfTest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8366421
+ * @summary Test for ModifiedUtf.utfLen() return type change from int to long to avoid overflow
+ * @modules java.base/jdk.internal.classfile.impl
+ *          java.base/jdk.internal.util
+ * @run main/othervm -Xmx4g --add-opens java.base/jdk.internal.classfile.impl=ALL-UNNAMED ModifiedUtfTest
```
Comment (liach) [inline test/jdk/jdk/internal/util/ModifiedUtfTest.java]:
```suggestion
 * @modules java.base/jdk.internal.classfile.impl:+open
 *          java.base/jdk.internal.util
 * @run main/othervm -Xmx4g ModifiedUtfTest
```

File: src/java.base/share/classes/java/io/ObjectOutputStream.java [hunk 6]
```diff
@@ -1899,12 +1901,12 @@ public void writeUTF(String str) throws IOException {
...
             int countNonZeroAscii = JLA.countNonZeroAscii(str);
-            int utflen = utfLen(str, countNonZeroAscii);
-            if (utflen <= 0xFFFF) {
+            long utflen = utfLen(str, countNonZeroAscii);
+            if (ExactConversionsSupport.isLongToCharExact(utflen)) {
                 if(writeHeader) {
                     writeByte(TC_STRING);
                 }
-                writeShort(utflen);
+                writeShort((int)utflen);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/io/ObjectOutputStream.java]: This would seem more consistent if the cast was to (short).

File: test/jdk/jdk/internal/util/ModifiedUtfTest.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+public class ModifiedUtfTest {
+
+  /**
+   * Keep only a fixed-length output and stop writing further data
+   * by throwing an exception when the limit is exceeded.
+   * For testing purposes only.
+   */
+  static class HeaderCaptureOutputStream extends OutputStream {
+    private byte[] head;
```
Comment (RogerRiggs) [inline test/jdk/jdk/internal/util/ModifiedUtfTest.java]: Indentation of java code is 4 spaces in OpenJDK.

**Comment by hgqxjj** [comment]
@RogerRiggs @liach Thanks for the review. I've updated the change as requested. Please take a look.

**Comment by hgqxjj** [comment]
@liach Thank you for the approval. I combined these tests into a single class mainly to reuse the large string.

@RogerRiggs  Do you have any further suggestions? If not, I’ll proceed with integrating it.

**Review by RogerRiggs — APPROVED** [review]
Looks good.  Thanks for the clever HeaderCaptureOutputStream test.

**Comment by liach** [comment]
This is in tier 1 so change in DOS/OOS/BufWriter that breaks this test should be relatively easy to catch.

**Comment by hgqxjj** [comment]
@RogerRiggs @liach 
Thanks for the approval and suggestion for this change.
I have integrated it. Could you please sponsor it? Thanks!

**Comment by liach** [comment]
I pulled and tested java/io, jdk/internal/util, jdk/classfile and things look fine, I assume this is good. My work at #27281 also depends on this change here.

/sponsor

**Comment by hgqxjj** [comment]
@liach Thanks！
